File: daily/2026/1559_Detect_Cycles_in_2D_Grid.cpp
#include <iostream>
#include <vector>

using namespace std;

class DSU {
public:
    vector<int> parents;
    DSU(int n){
        parents.resize(n);
        for (int i = 0; i < n; i++) parents[i] = i;
    }
    int find(int i){
        return parents[i] == i? i : parents[i] = find(parents[i]);
    }
    void combine(int i, int j){
        int x = find(i);
        int y = find(j);
        parents[y] = x;
    }
};

class Solution {
public:
    bool containsCycle(vector<vector<char>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        DSU u(m * n);
        for (int i = 0; i < m; i++){
            for (int j = 0; j < n; j++){
                // down
                if (i + 1 < m && grid[i][j] == grid[i+1][j]){
                    u.combine(i * n + j, (i + 1) * n + j);
                }
                // right
                if (j + 1 < n && grid[i][j] == grid[i][j+1]){
                    if (u.find(i * n + j) == u.find(i * n + j + 1)) return true;
                    u.combine(i * n + j, i * n + j + 1);
                }
            }
        }
        return false;
    }
};

int main(){
    vector<vector<char>> grid = {{'a','a','a','a'},{'a','b','b','a'},{'a','b','b','a'},{'a','a','a','a'}};
    // vector<vector<char>> grid = {{'c','c','c','a'},{'c','d','c','c'},{'c','c','e','c'},{'f','c','c','c'}};
    // vector<vector<char>> grid = {{'a','b','b'},{'b','z','b'},{'b','b','a'}};
    Solution S;
    cout << S.containsCycle(grid) << endl;
    return 0;
}

File: daily_python/1559_Detect_Cycles_in_2D_Grid.py
import os
import sys
from typing import List

class DSU:
    def __init__(self, n: int):
        self.parents = [i for i in range(n)]
    def find(self, i: int) -> int:
        if self.parents[i] != i:
            self.parents[i] = self.find(self.parents[i])
        return self.parents[i]
    def combine(self, i: int, j: int) -> None:
        x = self.find(i)
        y = self.find(j)
        self.parents[y] = x
        return
    
class Solution:
    def containsCycle(self, grid: List[List[str]]) -> bool:
        m, n = len(grid), len(grid[0])
        u = DSU(m * n)
        for i in range(m):
            for j in range(n):
                if i + 1 < m and grid[i][j] == grid[i+1][j]:
                    u.combine(i * n + j, (i + 1) * n + j)
                if j + 1 < n and grid[i][j] == grid[i][j+1]:
                    if u.find(i * n + j) == u.find(i * n + j + 1):
                        return True
                    u.combine(i * n + j, i * n + j + 1)
        return False
    
def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            grid = [list(row.split("','")) for row in lines[0].split('=')[1].strip()[3:-4].split("'},{'")]
            dataset.append(grid)
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for grid in dataset:
        results.append(solution.containsCycle(grid))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')